include/gridtools/stencil_composition/structured_grids/backend_mc/iterate_domain_mc.hpp

```diff
@@ -42,106 +39,77 @@ namespace gridtools {
+            }
         };
 
+#ifdef __SSE__
+        template <class T, class PtrDiff, class Strides>
+        GT_FORCE_INLINE void do_prefetch(T *ptr, PtrDiff offset, Strides const &strides, int_t dist) {
+            if (!dist)
+                return;
+            sid::shift(offset, sid::get_stride<dim::k>(strides), dist);
+            _mm_prefetch(reinterpret_cast<const char *>(ptr + offset), _MM_HINT_T1);
+        }
+
+        GT_FORCE_INLINE void do_prefetch(...) {}
+#endif
+
     } // namespace iterate_domain_mc_impl_
 
     /**
      * @brief Iterate domain class for the MC backend.
      */
-    template <typename IterateDomainArguments>
+    template <class LocalDomain, class IJCachedArgs>
     class iterate_domain_mc {
-        GT_STATIC_ASSERT(is_iterate_domain_arguments<IterateDomainArguments>::value, GT_INTERNAL_ERROR);
-
-        using local_domain_t = typename IterateDomainArguments::local_domain_t;
-        GT_STATIC_ASSERT(is_local_domain<local_domain_t>::value, GT_INTERNAL_ERROR);
-
-        using esf_sequence_t = typename IterateDomainArguments::esf_sequence_t;
-        using cache_sequence_t = typename IterateDomainArguments::cache_sequence_t;
-
-        /* meta function to get storage info index in local domain */
-        template <typename StorageInfo>
-        using local_domain_storage_index = meta::st_position<typename local_domain_t::strides_kinds_t, StorageInfo>;
-
-        /* meta function to check if a storage info belongs to a temporary field */
-        template <typename StorageInfo>
-        GT_META_DEFINE_ALIAS(
-            storage_is_tmp, meta::st_contains, (typename local_domain_t::tmp_strides_kinds_t, StorageInfo));
-
-        using ij_cache_args_t = GT_META_CALL(ij_cache_args, typename IterateDomainArguments::cache_sequence_t);
+        GT_STATIC_ASSERT(is_local_domain<LocalDomain>::value, GT_INTERNAL_ERROR);
 
-      public:
-        // the number of different storage metadatas used in the current functor
-        static const uint_t n_meta_storages = meta::length<typename local_domain_t::strides_kinds_t>::value;
-
-        using array_index_t = array<int_t, n_meta_storages>;
-        // *************** end of type definitions **************
-
-      private:
-        // *********************** members **********************
-        local_domain_t const &local_domain;
+        typename LocalDomain::strides_map_t const &m_strides_map;
+        typename LocalDomain::ptr_map_t m_ptr_map;
         int_t m_i_block_index;     /** Local i-index inside block. */
         int_t m_j_block_index;     /** Local j-index inside block. */
         int_t m_k_block_index;     /** Local/global k-index (no blocking along k-axis). */
         int_t m_i_block_base;      /** Global block start index along i-axis. */
         int_t m_j_block_base;      /** Global block start index along j-axis. */
         int_t m_prefetch_distance; /** Prefetching distance along k-axis, zero means no software prefetching. */
```

Comment:
`m_prefetch_distance` can be removed, too.

Reply:
done